src/NRedisStack/RedisUriParser.cs

```diff
@@ -0,0 +1,134 @@
+﻿using System.Runtime.CompilerServices;
+using System.Text.RegularExpressions;
+using StackExchange.Redis;
+
+[assembly: InternalsVisibleTo("NRedisStack.Tests")]
+
+namespace NRedisStack
+{
+    /// <summary>
+    /// URI parsing utility.
+    /// </summary>
+    internal static class RedisUriParser
+    {
+        /// <summary>
+        /// Parses a Config options for StackExchange Redis from the URI.
+        /// </summary>
+        /// <param name="redisUri">Redis Uri string</param>
+        /// <returns>A configuration options result for SE.Redis.</returns>
+        internal static ConfigurationOptions FromUri(string redisUri)
+        {
+            var options = new ConfigurationOptions();
+
+            if (string.IsNullOrEmpty(redisUri))
+            {
+                options.EndPoints.Add("localhost:6379");
+                return options;
+            }
+
+            var uri = new Uri(redisUri);
+            ParseHost(options, uri);
+            ParseUserInfo(options, uri);
+            ParseQueryArguments(options, uri);
+            ParseDefaultDatabase(options, uri);
+            options.Ssl = uri.Scheme == "rediss";
+            options.AbortOnConnectFail = false;
+            return options;
+        }
+
+        private static void ParseDefaultDatabase(ConfigurationOptions options, Uri uri)
+        {
+            if (string.IsNullOrEmpty(uri.AbsolutePath))
+            {
+                return;
+            }
+
+            var dbNumStr = Regex.Match(uri.AbsolutePath, "[0-9]+").Value;
+            int dbNum;
+            if (int.TryParse(dbNumStr, out dbNum))
+            {
+                options.DefaultDatabase = dbNum;
+            }
+        }
+
+        private static IList<KeyValuePair<string, string>> ParseQuery(string query) =>
+            query.Split('&').Select(x =>
+                new KeyValuePair<string, string>(x.Split('=').First(), x.Split('=').Last())).ToList();
+
+        private static void ParseUserInfo(ConfigurationOptions options, Uri uri)
+        {
+            if (!string.IsNullOrEmpty(uri.UserInfo))
+            {
+                var userInfo = uri.UserInfo.Split(':');
+                if (userInfo.Length > 1)
```

Comment:
Can we throw the exception on the if, and then continue in the fuction if not?

ie:
`if (userInfo.Length < 1) && userInfo.Length > 2                 {
                throw new FormatException(
                        "Username and password must be in the form username:password - if there is no username use the format :password");
                }
`

# do the options.User stuff